### ISSUE TITLE
Disambiguate what copy=True means for dask

### ISSUE THREAD

**Issue by crusaderky** [data-apis/array-api#866]
The current documentation for the `copy` parameter of `asarray` states:
https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html#asarray

> copy (Optional[bool]) – boolean indicating whether or not to copy the input. If True, the function must always copy. If False, the function must never copy for input which supports the buffer protocol and must raise a ValueError in case a copy would be necessary. If None, the function must reuse existing memory buffer if possible and copy otherwise. Default: None.

The meaning of `copy=True` is ambiguous for dask.
There are two possible interpretations:
1. updating the output array will never alter the input array; OR
2. the output array will never share memory with the input array; in other words dereferencing the input array can release its memory even if you still hold a reference to the output array.

In dask, updating a collection actually creates brand new graph nodes under the hood and repoints the collection to those nodes, so the original is never modified. However, the original chunks, generated e.g. by `from_array`, are still held inside the graph.

I strongly prefer the first definition, as IMHO decisions around memory management should be considered low level and delegated to the individual libraries.

**Comment by rgommers** [comment]
> I strongly prefer the first definition, as IMHO decisions around memory management should be considered low level and delegated to the individual libraries.

Yes agreed. In all such cases we've cared about the semantics, not about physical memory layout. A library should always be able to reuse memory or other such optimizations if it guarantees that that does not affect the semantics.

I'm pretty that using definition (1) is not controversial, and I remember clarifying this in some other place in the docs already - just can't remember where.

**Comment by ev-br** [comment]
If anything, it'd be quite helpful to clarify what copy=None means for dask, https://github.com/data-apis/array-api-compat/issues/209

**Comment by crusaderky** [comment]
> If anything, it'd be quite helpful to clarify what copy=None means for dask, [data-apis/array-api-compat#209](https://github.com/data-apis/array-api-compat/issues/209)

We could change "if possible" to "if possible and sensible"...

**Comment by asmeurer** [comment]
> I'm pretty that using definition (1) is not controversial, and I remember clarifying this in some other place in the docs already - just can't remember where.

https://github.com/data-apis/array-api/issues/788#issuecomment-2080416220

**Comment by rgommers** [comment]
Thanks for the link @asmeurer!

It seems like this issue overlaps a lot with that one. The "logical semantics" point stands I think - the difference is that for JAX copies sometimes do matter, while for Dask they never matter according to @crusaderky. So I think for Dask it's fine to not copy memory - would you agree?

The "never" here is of course the point of interest - it is probably possible to write some code where that never isn't true, especially if one starts mixing Dask and NumPy, which is fairly common.

**Comment by crusaderky** [comment]
xref https://github.com/data-apis/array-api/issues/867
